Feature release/hydrator 1323 add error flattener

albertshau opened the pull request.

albertshau committed:
HYDRATOR-1323 add an error flattener

The ErrorFlattener simply adds fields to all error records for
the error message, code, and stage.

albertshau commented:
should not be merged until https://github.com/caskdata/cdap/pull/7969 is merged

elfenheart commented:
@albertshau just a thought, would it be possible to add a getSchema method for this plugin, so that the user can propagate the correct schema?

albertshau commented:
ok can add that

albertshau committed:
HYDRATOR-1323 fix javascript emit error bug

javascript transform was assuming error records have the same
schema as the output, when it actually has the same schema as
the input.

albertshau committed:
HYDRATOR-1323 rename flattener to collector

anew commented on core-plugins/docs/ErrorCollector-errortransform.md:
```diff
@@ -0,0 +1,61 @@
+Properties
+----------
+**messageField:** The name of the error message field to use in the output schema. Defaults to 'errMsg'.
+If this is not specified, the error message will be dropped.
```
Pick one: Does it default to errMsg, or will the messages be dropped?

albertshau replied:
in the UI, it has a default value of 'errMsg'. If the user chooses to delete it, the message will be dropped. I can probably make this clearer.

anew commented on core-plugins/docs/ErrorCollector-errortransform.md:
```diff
@@ -0,0 +1,61 @@
+
+**codeField:** The name of the error code field to use in the output schema. Defaults to 'errCode'.
+If this is not specified, the error code will be dropped.
+
```
same here

anew commented on core-plugins/docs/ErrorCollector-errortransform.md:
```diff
@@ -0,0 +1,61 @@
+
+**stageField:** The name of the error stage field to use in the output schema. Defaults to 'errStage'.
+If this is not specified, the error stage will be dropped.
+
```
and here

anew commented on core-plugins/src/main/java/co/cask/hydrator/plugin/error/ErrorCollector.java:
```diff
@@ -0,0 +1,144 @@
+import javax.ws.rs.Path;
+
+/**
+ * Adds the error code and error message to each record, then emits it.
```
and error stage?

anew commented on core-plugins/src/main/java/co/cask/hydrator/plugin/error/ErrorCollector.java:
```diff
@@ -0,0 +1,144 @@
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    if (inputSchema != null) {
+      if (inputSchema.getField(config.messageField) != null) {
```
can this produce NPE if config.messageField is not set? and the same for the next two

anew reviewed (commented):
I left a few comments. Can't really speak to the widget, but it seems that @elfenheart has taken a look at that

albertshau committed:
HYDRATOR-1323 protect against NPE and clarify docs

anew reviewed (approved):
LGTM
